components/brave_new_tab/resources/components/settings/background_panel.tsx

```
@@ -0,0 +1,193 @@
+    state.selectedBackgroundType,
+    state.selectedBackground,
+    state.braveBackgrounds,
+    state.customBackgrounds
+  ])
+
+  const [panelType, setPanelType] = React.useState<BackgroundType>('none')
+  const [uploading, setUploading] = React.useState(false)
+
+  function getTypePreviewValue(type: BackgroundType) {
+    const isSelectedType = type === selectedBackgroundType
+    switch (type) {
+      case 'brave':
+        return braveBackgrounds.at(0)?.imageUrl || ''
+      case 'custom':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return customBackgrounds.at(0) ?? ''
+      case 'solid':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return solidPreviewBackground
+      case 'gradient':
+        if (isSelectedType && selectedBackground) {
+          return selectedBackground
+        }
+        return gradientPreviewBackground
+      case 'none':
+        return ''
+    }
+  }
+
+  function renderUploadPreview() {
+    return (
+      <div className='preview upload'>
+        {uploading ? <ProgressRing /> : <Icon name='upload' />}
+        {getString('uploadBackgroundLabel')}
+      </div>
+    )
+  }
+
+  function renderTypePreview(type: BackgroundType) {
+    if (type === 'custom' && customBackgrounds.length === 0) {
+      return renderUploadPreview()
+    }
+    return (
+      <div
+        className='preview'
+        style={inlineCSSVars({
+          '--preview-background':
+            backgroundCSSValue(type, getTypePreviewValue(type))
+        })}
+      >
+        {
+          type === selectedBackgroundType &&
+            <span className='selected-marker'>
+              <Icon name='check-normal' />
+            </span>
+        }
+      </div>
+    )
+  }
+
+  function showCustomBackgroundChooser() {
+    model.showCustomBackgroundChooser().then(wrapCallback((selected) => {
+      if (selected && !uploading) {
```

won't this just silently fail if it's currently uploading?

---

The UI should not allow the user to get into a state where they are trying to upload an image while an image is currently being uploaded (it will show a spinner instead of the upload button). But if for some reason it did happen, we want to avoid having two uploads in progress and further confusing state.

---

:+1: to this - seems like a pretty reasonable approach